File: Algorithm_380_Insert Delete GetRandom O(1).cpp
#include <iostream>
#include <unordered_map>
using namespace std;
class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if (m.find(val) != m.end()) return false;
        nums.emplace_back(val);
        m[val] = nums.size() - 1; // m stores the index of val in nums
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if (m.find(val) == m.end()) return false;
        int last = nums.back(); // the latest inserted value
        m[last] = m[val]; // replace the index of last with that of val
        nums[m[val]] = last; // replace val with last, which is identical to erase val from nums
        nums.pop_back();
        m.erase(val); // erase the index of val in m
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        return nums[rand() % nums.size()];
    }
private:
    vector<int> nums;
    unordered_map<int, int> m;
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * bool param_1 = obj.insert(val);
 * bool param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: Algorithm_380_Insert Delete GetRandom O(1).java
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

// 思路：用map存储val到这个val在list中的下标的映射
// 插入时，直接插入到数组的尾部，并在map中创建相应的entry
// 查询时，直接通过随机数产生一个在list的范围内的下标，并返回相应的值
// 删除时，先通过map获取val对应的下标位置
// 如果下标恰好为list.size() - 1，则直接把val从map和list中删除，都是O(1)
// 若不是list.size() - 1，则先把list中的最后一个元素交换到当前位置上
// 最后更新map中这个元素对应的entry，即改变相应的list中的下标
// 最后从list中删除最后一个元素，并从map中删除val对应的entry

class RandomizedSet {

    private List<Integer> list;
    private Map<Integer, Integer> map;

    /** Initialize your data structure here. */
    public RandomizedSet() {
        list = new ArrayList<>();
        map = new HashMap<>();
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    public boolean insert(int val) {
        if (map.containsKey(val)) {
            return false;
        }
        list.add(val);
        map.put(val, list.size() - 1);
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    public boolean remove(int val) {
        if (!map.containsKey(val)) {
            return false;
        }
        int index = map.get(val);
        if (index != list.size() - 1) {
            map.put(list.get(list.size() - 1), index);
            list.set(index, list.get(list.size() - 1));
        }
        map.remove(val);
        list.remove(list.size() - 1);
        return true;
    }
    
    /** Get a random element from the set. */
    public int getRandom() {
        Random r = new Random();
        return list.get(r.nextInt(list.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */